src/builtins/OWNERSHIP.md

```diff
@@ -0,0 +1,225 @@
+# Ownership Semantics in Roc Builtins
+
+This document defines the canonical terminology for ownership semantics in Roc's builtin functions.
+Understanding these patterns is critical for correctly implementing and calling builtins.
+
+## Core Invariant
+
+**refcount = number of live references to the data**
+
+When refcount reaches 0, memory is freed.
+
+Basic operations:
+1. **Create**: allocate with refcount = 1
+2. **Share**: increment refcount (data is shared, both references valid)
+3. **Release**: decrement refcount (if 0, free memory)
+
+---
+
+## Argument Handling (2 patterns)
+
+These describe how a function treats its input arguments.
+
+### Borrow
+
+A **borrowing** function reads its arguments without affecting their refcount.
+
+- Caller retains ownership
+- No refcount change at call boundary
+- Caller can still use argument after call
+
+**Examples**: `strEqual`, `listLen`, `strContains`, `countUtf8Bytes`
+
+### Consume
+
+A **consuming** function takes ownership of its argument.
+
+- Caller transfers ownership to callee
+- Caller must not use argument after call (logically moved)
+- Function is responsible for cleanup (decref when done)
+
+**Examples**: `strConcat`, `listConcat`, `strJoinWith`
+
+---
+
+## Result Patterns (3 types)
+
+These describe the relationship between a function's result and its arguments.
+
+### Independent
+
+Result is a new allocation, unrelated to arguments.
+
+- Normal ownership: caller owns result
+- Caller must decref when done
+
+**Example**: `strConcat` returns newly allocated combined string
+
+### Copy-on-Write (Same-if-unique)
+
+Result may be the same allocation as an argument.
+
+- Consumes the input argument
+- If `isUnique()`: mutates in place, returns same pointer
+- If shared: decrefs argument internally, allocates new, returns new pointer
+- **Critical**: `result.ptr` may equal `arg.ptr`
+
+**Examples**: `strWithAsciiUppercased`, `strTrim`, `listAppend`
+
+**Interpreter handling**: Check if `result.bytes == arg.bytes`:
```

Comment:
This does not feel right to me. For example, bytes could have changed due to a call to `realloc` to resize the underlying string. Likely this is an interpreter bug if that is what the code is checking.